src/Architectures.jl

```diff
@@ -54,46 +66,61 @@ On single-process, non-distributed systems, return `arch`.
 child_architecture(arch) = arch
 
 array_type(::CPU) = Array
-array_type(::GPU) = CuArray
+array_type(::CUDAGPU) = CuArray
+array_type(::ROCmGPU) = ROCArray
 
 arch_array(::CPU, a::Array)   = a
 arch_array(::CPU, a::CuArray) = Array(a)
-arch_array(::GPU, a::Array)   = CuArray(a)
-arch_array(::GPU, a::CuArray) = a
-
+arch_array(::CPU, a::ROCArray) = Array(a)
+arch_array(::CUDAGPU, a::Array) = CuArray(a)
+arch_array(::ROCmGPU, a::Array) = ROCArray(a)
+arch_array(::CUDAGPU, a::CuArray) = a
+arch_array(::ROCmGPU, a::ROCArray) = a
 arch_array(::CPU, a::BitArray) = a
-arch_array(::GPU, a::BitArray) = CuArray(a)
+arch_array(::CUDAGPU, a::BitArray) = CuArray(a)
+arch_array(::ROCmGPU, a::BitArray) = ROCArray(a)
+
+arch_array(::CUDAGPU, a::SubArray{<:Any, <:Any, <:CuArray}) = a
+arch_array(::ROCmGPU, a::SubArray{<:Any, <:Any, <:ROCArray}) = a
```

Comment:
these two lines are the same
```suggestion
arch_array(::GPU{D}, a::SubArray{<:Any, <:Any, <:CuArray}) where D = a
```